[Style] Convert border-image-* and mask-border-* to strong style types

weinig opened the pull request:
#### 8c1ac181ea22e32b72c3a680f3faf3756958eb14
<pre>
[Style] Convert border-image-* and mask-border-* to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=296862">https://bugs.webkit.org/show_bug.cgi?id=296862</a>

Reviewed by Antti Koivisto.

Converts the border-image-* and mask-border-* family of CSS properties to
use strong style types, replacing `NinePieceImage` with separate
`Style::BorderImage` and `Style::MaskBorder` types. Painting stays shared
via a new `NinePieceImagePainter`.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/CSSToStyleMap.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/rendering/BorderPainter.cpp:
* Source/WebCore/rendering/BorderPainter.h:
* Source/WebCore/rendering/InlineBoxPainter.cpp:
* Source/WebCore/rendering/NinePieceImagePainter.cpp: Added.
* Source/WebCore/rendering/NinePieceImagePainter.h: Added.
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/NinePieceImage.cpp: Removed.
* Source/WebCore/rendering/style/NinePieceImage.h: Removed.
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleReflection.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImage.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageOutset.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageOutset.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageRepeat.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageRepeat.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageSlice.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageSlice.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageSource.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageSource.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h: Added.
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
* Source/WebCore/style/values/color/StyleColorMix.cpp:
* Source/WebCore/style/values/images/StyleImageWrapper.cpp:
* Source/WebCore/style/values/images/StyleImageWrapper.h:
* Source/WebCore/style/values/masking/StyleMaskBorder.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorder.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderOutset.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderOutset.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderRepeat.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderRepeat.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderSlice.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderSlice.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderSource.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderSource.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderWidth.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderWidth.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:

Canonical link: <a href="https://commits.webkit.org/298261@main">https://commits.webkit.org/298261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e4a5ae30632f89bbc4d15a6989cc3cfaba24e15

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114553 "27 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24762 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120717 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65278 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34928 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42859 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87066 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41975 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117501 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27833 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102873 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67457 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27018 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21000 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64399 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97211 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21113 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123924 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41566 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31026 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95886 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41943 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99063 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95669 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24411 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40833 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18669 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37630 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41445 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46955 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41031 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44338 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42781 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/values/masking/StyleMaskBorderSlice.h:
```diff
@@ -0,0 +1,106 @@
+namespace Style {
+
+// <mask-border-slice-value> = <number [0,∞]> | <percentage [0,∞]>
+struct MaskBorderSliceValue {
```
The value types are pretty regular. Do you have plans to autogenerate them at some point?

weinig replied:
The current goal is to simplify these down to just a single simple declaration when my LengthWrapperBase changes all complete.

weinig committed:
[Style] Convert border-image-* and mask-border-* to strong style types
https://bugs.webkit.org/show_bug.cgi?id=296862

Reviewed by Antti Koivisto.

Converts the border-image-* and mask-border-* family of CSS properties to
use strong style types, replacing `NinePieceImage` with separate
`Style::BorderImage` and `Style::MaskBorder` types. Painting stays shared
via a new `NinePieceImagePainter`.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/CSSToStyleMap.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/rendering/BorderPainter.cpp:
* Source/WebCore/rendering/BorderPainter.h:
* Source/WebCore/rendering/InlineBoxPainter.cpp:
* Source/WebCore/rendering/NinePieceImagePainter.cpp: Added.
* Source/WebCore/rendering/NinePieceImagePainter.h: Added.
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/NinePieceImage.cpp: Removed.
* Source/WebCore/rendering/style/NinePieceImage.h: Removed.
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleReflection.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImage.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageOutset.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageOutset.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageRepeat.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageRepeat.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageSlice.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageSlice.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageSource.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageSource.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h: Added.
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
* Source/WebCore/style/values/color/StyleColorMix.cpp:
* Source/WebCore/style/values/images/StyleImageWrapper.cpp:
* Source/WebCore/style/values/images/StyleImageWrapper.h:
* Source/WebCore/style/values/masking/StyleMaskBorder.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorder.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderOutset.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderOutset.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderRepeat.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderRepeat.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderSlice.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderSlice.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderSource.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderSource.h: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderWidth.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskBorderWidth.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:

Canonical link: https://commits.webkit.org/298261@main
